packages/yarnpkg-core/sources/Project.ts

```diff
@@ -1004,23 +1010,36 @@ export class Project {
         try {
           installStatus = await installer.installPackage(pkg, fetchResult);
         } finally {
-          if (fetchResult.releaseFs) {
-            fetchResult.releaseFs();
+          // We don't close the archive if the linker is still installing things; it
+          // will be closed once its install promise resolves
+          if (!installStatus?.installPromise) {
+            fetchResult.releaseFs?.();
```

Comment:
Could we run multiple `installPackage`s at the same time instead of adding this?

Reply:
I'd prefer not - if we did, linkers would lose their deterministic execution order, which could lead to subtle differences in the generated artifacts depending on each linker's implementation (could be from lines not in the same order in the PnP file to different hoisting due to a different traversal).

So instead, the `installPromise` lets linker define a concurrent optional execution mode while preserving the deterministic execution.

Reply:
In that case I would prefer if the `pnpm` linker handles this internally in `finalizeInstall` like the `node-modules` linker does.

Reply:
The NM linker needs to have the full dependency tree before being able to start processing it. By contrast the pnpm linker can generate the artifacts as a stream.

Perhaps an alternative would be a mode to explicitly opt-out of the deterministic execution mode.

Reply:
Some sorting in `attachInternalDependencies`, `attachExternalDependents`, and `finalizeInstall` should make the `install-state` identical at least.

Reply:
I refactored the code. Now installers are responsible of managing their own queue if they need it. It's not that much extra complexity, and avoids overloading the core 👍